src/query/block/builder.go

```diff
@@ -0,0 +1,156 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package block
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/m3db/m3/src/query/cost"
+	"github.com/m3db/m3/src/query/models"
+	xcost "github.com/m3db/m3/src/x/cost"
+
+	"github.com/uber-go/tally"
+)
+
+// ColumnBlockBuilder builds a block optimized for column iteration.
+type ColumnBlockBuilder struct {
+	block           *columnBlock
+	enforcer        cost.ChainedEnforcer
+	blockDatapoints tally.Counter
+}
+
+// NewColumnBlockBuilder creates a new column block builder.
+func NewColumnBlockBuilder(
+	queryCtx *models.QueryContext,
+	meta Metadata,
+	seriesMeta []SeriesMeta) Builder {
+	return ColumnBlockBuilder{
+		enforcer: queryCtx.Enforcer.Child(cost.BlockLevel),
+		blockDatapoints: queryCtx.Scope.Tagged(
+			map[string]string{"type": "generated"}).Counter("datapoints"),
+		block: &columnBlock{
+			meta:       meta,
+			seriesMeta: seriesMeta,
+			blockType:  BlockDecompressed,
+		},
+	}
+}
+
+// AppendValue adds a value to a column at index.
+func (cb ColumnBlockBuilder) AppendValue(idx int, value float64) error {
+	columns := cb.block.columns
+	if len(columns) <= idx {
+		return fmt.Errorf("idx out of range for append: %d", idx)
+	}
+
+	r := cb.enforcer.Add(1)
+	if r.Error != nil {
+		return r.Error
+	}
+
+	cb.blockDatapoints.Inc(1)
+
+	columns[idx] = append(columns[idx], value)
+	return nil
+}
+
+// AppendValues adds a slice of values to a column at index.
+func (cb ColumnBlockBuilder) AppendValues(idx int, values []float64) error {
+	columns := cb.block.columns
+	if len(columns) <= idx {
+		return fmt.Errorf("idx out of range for append: %d", idx)
+	}
+
+	r := cb.enforcer.Add(xcost.Cost(len(values)))
+	if r.Error != nil {
+		return r.Error
+	}
+
+	cb.blockDatapoints.Inc(int64(len(values)))
+	columns[idx] = append(columns[idx], values...)
+	return nil
+}
+
+// AddCols adds the given number of columns to the block.
+func (cb ColumnBlockBuilder) AddCols(num int) error {
+	if num < 1 {
+		return fmt.Errorf("must add more than 0 columns, adding: %d", num)
+	}
+
+	newCols := make([]column, num)
+	cb.block.columns = append(cb.block.columns, newCols...)
+	return nil
+}
+
+// PopulateColumns sets all columns to the given row size.
+func (cb ColumnBlockBuilder) PopulateColumns(size int) {
+	cols := make([]float64, size*len(cb.block.columns))
+	for i := range cb.block.columns {
+		cb.block.columns[i] = cols[size*i : size*(i+1)]
+	}
+}
+
+// SetRow sets a given block row to the given values and metadata.
+func (cb ColumnBlockBuilder) SetRow(
+	idx int,
+	values []float64,
+	meta SeriesMeta,
+) error {
+	cols := cb.block.columns
+	if len(values) == 0 {
+		// Sanity check. Should never happen.
+		return errors.New("cannot insert empty values")
```

Comment:
nit: Make this error at top of file?

Reply:
Any general guides on when we do error at the top vs inline?

Reply:
The current general guidelines tends to be (and we should really put this in CI instead of just informal guidelines):
- If requires extra data to be put in the error message, then need to use fmt.Errorf(...) or similar so must allocate
- If does not require any data and the error is effectively constant, then having a single shared instance of the error is desirable